castle/cms/indexing/hps.py

```diff
@@ -0,0 +1,335 @@
+from wildcard.hps.opensearch import WildcardHPSCatalog
+from zope.component import getMultiAdapter
+from zope.component import getUtility
+
+
+logger = logging.getLogger("Plone")
+
+
+def get_catalog():
+    catalog = api.portal.get_tool('portal_catalog')
+    hps = WildcardHPSCatalog(catalog)
+    return hps
+
+
+def get_connection():
+    hpscat = get_catalog()
+    return hpscat.connection
+
+
+def get_index_name():
+    hpscat = get_catalog()
+    return hpscat.index_name
+
+
+def get_bulk_size():
+    hpscat = get_catalog()
+    return hpscat.get_setting("bulk_size", 50)
+
+
+def is_enabled():
+    hpscat = get_catalog()
+    return hpscat.enabled
+
+
+def health_is_good():
+    conn = get_connection()
+    try:
+        return conn.cluster.health()['status'] in ('green', 'yellow')
+    except Exception:
+        return False
+
+
+def hps_get_number_of_matches(index_name, query):
+    conn = get_connection()
+    results = conn.count(index=index_name, body=query)
+    return results.get('count', -1)
+
+
+def hps_get_data(index_name, query, **kwargs):
+    conn = get_connection()
+    results = conn.search(index=index_name, body=query, **kwargs)
+    try:
+        data = results['hits']['hits']
+        total = results['hits']['total']['value']
+        scroll_id = results.get('_scroll_id', None)
+        return data, total, scroll_id
+    except KeyError:
+        return [], -1, None
+
+
+def hps_get_scroll(scroll_id, **kwargs):
+    conn = get_connection()
+    results = conn.scroll(scroll_id=scroll_id, **kwargs)
+    try:
+        data = results['hits']['hits']
+        total = results['hits']['total']['value']
+        scroll_id = results.get('_scroll_id', None)
+        return data, total, scroll_id
+    except KeyError:
+        return [], -1, None
+
+
+def gen_audit_query(
+        field_map_prefix=None, sort=None, sortdir=None,
+        schema_version="1", schema_type="castle.cms.audit", instance=None, site=None,
+        typeval=None, user=None, content=None, after=None, before=None):
+
+    sortfield = sort
+    schemaversionfield = "schema_version"
+    schematypefield = "schema_type"
+    instancefield = "instance"
+    sitefield = "site"
+    typefield = "type"
+    userfield = "user"
+    objectfield = "object"
+    datefield = "date"
+    if field_map_prefix is not None:
+        sortfield = "{}.{}".format(field_map_prefix, sortfield)
+        schemaversionfield = "{}.{}".format(field_map_prefix, schemaversionfield)
+        schematypefield = "{}.{}".format(field_map_prefix, schematypefield)
+        instancefield = "{}.{}".format(field_map_prefix, instancefield)
+        sitefield = "{}.{}".format(field_map_prefix, sitefield)
+        typefield = "{}.{}".format(field_map_prefix, typefield)
+        userfield = "{}.{}".format(field_map_prefix, userfield)
+        objectfield = "{}.{}".format(field_map_prefix, objectfield)
+        datefield = "{}.{}".format(field_map_prefix, datefield)
+
```

Comment:
agreed, could be -- choice is mostly aesthetic, though, so I'll keep it since it's the code that's been reviewed and doesn't have a known error in it.